File: lev1/숫자 문자열과 영단어.py
def solution(s):
    word2num = {
        "zero": "0",
        "one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9"
    }
    s = s.lower()
    N = len(s)
    ans = ""
    # 0.16021884605288506
    #idx = 0
    #while idx < N:
    #    if s[idx].isdigit():
    #        ans += s[idx]
    #        idx += 1
    #        for j in [3, 4, 5]:
    #            if s[idx:idx+j] in word2num:
    #                ans += word2num[s[idx:idx+j]]
    #                idx += j
    #                break
    #return int(ans)

    # code refactoring - 0.29748598858714104
    #start, end = 0, 1
    #while end <= N:
    #    if s[start:end].isdigit():
    #        ans += s[start:end]
    #        start += 1
    #    elif s[start:end] in word2num:
    #        ans += word2num[s[start:end]]
    #        start = end
    #    end += 1
    #return int(ans) 

    # code refactoring (R) - 0.0979218315333128
    for k, v in word2num.items():
        s = s.replace(k, v)
    return int(s)


print(solution("one4seveneight"))
print(solution("23four5six7"))
print(solution("2three45sixseven"))
print(solution("1234567891"))
print(solution("onetwothreefourfivesixseveneightnineone"))


if __name__ == "__main__":
    from timeit import Timer
    query = ["one4seveneight", "23four5six7", "2three45sixseven", "123", "onetwothreefourfivesixseveneightnineone"]
    t = Timer(f"for t in {query}: solution(t)", "from __main__ import solution")
    print(t.timeit(number=10000))

File: lev1/숫자_문자열과_영단어.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>

// 파라미터로 주어지는 문자열은 const로 주어집니다. 변경하려면 문자열을 복사해서 사용하세요.
int solution(const char* s) {
    char *table[] = {
        "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
    };
    int ans = 0;
    while(*s){
        if(isdigit(*s))
            ans = ans*10 + *s-'0';
        else{
            for(int j=0; j<10; j++){
                if(strncmp(s, table[j], strlen(table[j])) == 0){
                    ans = ans*10 + j;
                    s += strlen(table[j])-1;
                }
            }
        }
        s++;
    }
    return ans;
}


int main(){
    printf("%d\n", solution("one4seveneight")); //1478
    printf("%d\n", solution("23four5six7")); //234567
    printf("%d\n", solution("2three45sixseven")); //234567
    printf("%d\n", solution("123")); //123
    printf("%d\n", solution("onetwothree")); //123
    return 0;
}
